pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,719 @@
+	// entityItems will be associated with the same labelItem if they have same Namespace, entityType, and labels.
+	labelKey string
+}
+
+// labelItem represents an individual label set. It's the actual object that will be matched with label selectors.
+// Entities of same type in same Namespace having same labels will share a labelItem.
+type labelItem struct {
+	// The label set that will be used for matching.
+	labels labels.Set
+	// The namespace of the entities that share the labelItem.
+	namespace string
+	// The type of the entities that share the labelItem.
+	entityType entityType
+	// The keys of the entityItems that share the labelItem.
+	entities sets.String
+	// The keys of the selectorItems that match the labelItem.
+	selectors sets.String
+}
+
+// groupItem contains a group's metadata and its selector.
+type groupItem struct {
+	// The type of the group.
+	groupType string
+	// The name of the group. It must be unique within its own type.
+	name string
+	// The selector of the group.
+	selector *types.GroupSelector
+	// selectorKey is the key of the selectorItem that the groupItem is associated with.
+	// groupItems will be associated with the same selectorItem if they have same selector.
+	selectorKey string
+}
+
+// selectorItem represents an individual label selector. It's the actual object that will be matched with label sets.
+// Groups having same label selector will share a selectorItem.
+type selectorItem struct {
+	// The label selector that will be used for matching.
+	selector *types.GroupSelector
+	// The keys of the groupItems that share the selectorItem.
+	groups sets.String
+	// The keys of the labelItems that match the selectorItem.
+	labels sets.String
+}
+
+var _ GroupEntityIndex = &groupEntityIndex{}
+
+// groupEntityIndex implements GroupEntityIndex.
+//
+// It abstracts label set and label selector from entities and groups and do actual matching against the formers to
+// avoid redundant calculation given that most entities (Pod or ExternalEntity) actually share labels. For example, Pods
+// that managed by a deployment controller have same labels.
+//
+// It maintains indexes between label set and label selector so that querying label sets that match a label selector and
+// reversed queries can be performed with a consistent time complexity. Indirectly, querying entities that match a group
+// and reversed queries can be performed with same complexity.
+//
+// The relationship of the four items are like below:
+// entityItem <===> labelItem <===> selectorItem <===> groupItem
+type groupEntityIndex struct {
+	lock sync.RWMutex
+
+	// entityItems stores all entityItems.
+	entityItems map[string]*entityItem
+
+	// labelItems stores all labelItems.
+	labelItems map[string]*labelItem
+	// labelIndex is nested map from entityType to Namespace to keys of labelItems.
+	// It's used to filter potential labelItems when matching a namespace scoped selectorItem.
+	labelIndex map[entityType]map[string]sets.String
```

Comment:
Is it more efficient to create two separate maps for Pods and ExternalEntities?

Reply:
Or to share code, we can convert them to one internal type?

Reply:
The code of matching Pods and ExtenalEntities is already shared and they are treated in same way when stored in labelItems.
Separating them in labelIndex is to reduce the potential labelItems a selector needs to match. This is because a selector can either select ExternalEntity or Pod. If it's a Pod selector, this is no need to try ExternalEntity labelItems. The same reason applies to selectorIndex.